### ISSUE TITLE
aws_security_group: AWS API summarizes rule specs, causing perpetual diffs and diff mismatches

### ISSUE THREAD

**Issue by clstokes** [hashicorp/terraform#2471]
With an `aws_security_group` resource with multiple ingress rules that use the same ports, Terraform will apply modifications each time an apply is executed even though no changes are needed. If the same rules are combined into a single ingress, this doesn't occur and if the rules have different ports this doesn't occur.

The use case for having multiple ingress with the same ports is to have different ingress definitions for different variables or blocks rather than a long and potentially confusing `${split(",",concat(var.block2,",",var.block2,",",var.block3))}` definition.

This is with Terraform v0.5.3. The resource below will cause this behavior. I would expect *no* modifications to be made from one `apply` to the next with this configuration, but terraform will re-apply the rules with each run.

A complete example with output is at https://gist.github.com/clstokes/4e8c7456f19a04ecf45d.

```
resource "aws_security_group" "causes_modifications_multiple_ingress_same_ports" {
  name = "causes_modifications_multiple_ingress_same_ports"
  description = "this security group CAUSES modifications each run"

  vpc_id = "vpc-abc123"

  ingress {
      from_port = 22
      to_port = 22
      protocol = "tcp"
      cidr_blocks = ["10.0.0.1/32"]
  }

  ingress {
      from_port = 22
      to_port = 22
      protocol = "tcp"
      cidr_blocks = ["10.0.0.2/32"]
  }

}
```

**Comment by ryanking** [comment]
I see a similar error with a single ingress block and multiple cidr blocks listed.

**Comment by phinze** [comment]
Thanks for the report and the testing you did to provide repro case, @clstokes - super helpful!

Nailed down the root cause, still haven't quite figured out how we should handle it.

Basically AWS is "summarizing" the SG rules based on the API requests we send.

Here are details from an `AuthorizeSecurityGroupIngress` we send - note two `IPPermission`s:

```
{
  Description: "this security group CAUSES modifications each run",
  GroupID: "sg-fbff3f9f",
  GroupName: "causes_modifications_multiple_ingress_same_ ports",
  OwnerID: "762203914556",
  VPCID: "vpc-3e66f85b"
} ingress rule: []*ec2.IPPermission{{
  FromPort: 22,
  IPProtocol: "tcp",
  IPRanges: [{
      CIDRIP: "10.0.0.2/32"
    }],
  ToPort: 22
}, {
  FromPort: 22,
  IPProtocol: "tcp",
  IPRanges: [{
      CIDRIP: "10.0.0.1/32"
    }],
  ToPort: 22
}}
```

AWS Responds by creating an SG that has _one_ `IPPermission` with multiple `IPRanges`:

```
{
  Description: "this security group CAUSES modifications each run",
  GroupID: "sg-fbff3f9f",
  GroupName: "causes_modifications_multiple_ingress_same_ ports",
  IPPermissions: [{
      FromPort: 22,
      IPProtocol: "tcp",
      IPRanges: [{
          CIDRIP: "10.0.0.2/32"
        },{
          CIDRIP: "10.0.0.1/32"
        }],
      ToPort: 22
    }],
  OwnerID: "762203914556",
  VPCID: "vpc-3e66f85b"
}
```

That means it will take more work to be able to properly distinguish between an `ingress` with multiple CIDR blocks and multiple ingresses.

I have a feeling the best solution here will be to dive into #2275 and make this nested block syntax spit out the equivalent of top level AWS SG rules. I'll follow up when I've had a chance to make progress there.

In the meantime, combining CIDR blocks will have to be the workaround for this issue.

@ryanking I'd be curious to see your example, since I can't think of a way for a single ingress block w/ multiple CIDRs to trigger this behavior.

**Comment by ryanking** [comment]
Looking more closely, I think I know what's going on (see below for configuration and output).

When I first created this security group it allowed 0.0.0.0/0. I then went in an replaced that with a list of CIDR blocks.

I then accidentally added a block with /0 (whoops), which amazon probably translated to 0.0.0.0/0.

After fixing the config and deleted that entry manually we're fine.

# Configuration
resource "aws_security_group" "bastion" {
  name = "${var.deploy_environment}-bastion"
  description = "Allow SSH traffic from the Internet"

  ingress {
    from_port = 22
    to_port = 22
    protocol = "tcp"

    cidr_blocks = [ "208.72.137.135/32",
                    "192.195.82.108/32",
                    "50.174.212.0/24",
                    "24.130.8.160/32",
                    "70.36.236.10/32",
                    "76.14.85.166/32",
                    "67.221.169.243/0"
		]
  }

  vpc_id = "${aws_vpc.vpc.id}"
}
# Plan output

~ aws_security_group.bastion
    ingress.1691870362.cidr_blocks.#:     "0" => "7"
    ingress.1691870362.cidr_blocks.0:     "" => "208.72.137.135/32"
    ingress.1691870362.cidr_blocks.1:     "" => "192.195.82.108/32"
    ingress.1691870362.cidr_blocks.2:     "" => "50.174.212.0/24"
    ingress.1691870362.cidr_blocks.3:     "" => "24.130.8.160/32"
    ingress.1691870362.cidr_blocks.4:     "" => "70.36.236.10/32"
    ingress.1691870362.cidr_blocks.5:     "" => "76.14.85.166/32"
    ingress.1691870362.cidr_blocks.6:     "" => "67.221.169.243/0"
    ingress.1691870362.from_port:         "" => "22"
    ingress.1691870362.protocol:          "" => "tcp"
    ingress.1691870362.security_groups.#: "0" => "0"
    ingress.1691870362.self:              "" => "0"
    ingress.1691870362.to_port:           "" => "22"
    ingress.732351028.cidr_blocks.#:      "7" => "0"
    ingress.732351028.cidr_blocks.0:      "208.72.137.135/32" => ""
    ingress.732351028.cidr_blocks.1:      "192.195.82.108/32" => ""
    ingress.732351028.cidr_blocks.2:      "50.174.212.0/24" => ""
    ingress.732351028.cidr_blocks.3:      "24.130.8.160/32" => ""
    ingress.732351028.cidr_blocks.4:      "70.36.236.10/32" => ""
    ingress.732351028.cidr_blocks.5:      "76.14.85.166/32" => ""
    ingress.732351028.cidr_blocks.6:      "0.0.0.0/0" => ""
    ingress.732351028.from_port:          "22" => "0"
    ingress.732351028.protocol:           "tcp" => ""
    ingress.732351028.security_groups.#:  "0" => "0"
    ingress.732351028.self:               "0" => "0"
    ingress.732351028.to_port:            "22" => "0"

**Comment by phinze** [comment]
@ryanking cool - glad you figured it out. we'll keep tracking the multiple ingress rules thing on this thread.

**Comment by phinze** [comment]
Joining the diff mismatch in #2692 to this issue since they both come from the same root cause and any fix we make will have to address both issues.

**Comment by catsby** [comment]
Hey all – this was fixed in #5100. 

Example:

```
resource "aws_vpc" "dummy" {
  cidr_block = "10.0.0.0/16"

  tags {
    Name = "test_dummy"
  }
}

resource "aws_security_group" "causes_modifications_multiple_ingress_same_ports" {
  name        = "causes_modifications_multiple_ingress_same_ports"
  description = "this security group CAUSES modifications each run"

  vpc_id = "${aws_vpc.dummy.id}"

  ingress {
    from_port   = 22
    to_port     = 22
    protocol    = "tcp"
    cidr_blocks = ["10.0.0.1/32"]
  }

  ingress {
    from_port   = 22
    to_port     = 22
    protocol    = "tcp"
    cidr_blocks = ["10.0.0.2/32"]
  }
}
```
no longer produces a diff

Thanks!

**Comment by joshuabaird** [comment]
@catsby I'm seeing this issue again in 0.10.3.  Is is it possible that this bug/issue has resurfaced?

**Comment by catsby** [comment]
@joshuabaird that's the same issue; AWS groups rules by `to_port`, `from_port`, and `protocol`. If you rules have those same values, then they need to be in the same stanza.

If you'd like to split them out into their own _things_, consider using `aws_security_group_rule` resources:

- https://www.terraform.io/docs/providers/aws/r/security_group_rule.html

**Comment by joshuabaird** [comment]
Yep - that is the workaround I am using now.  I was under the impression that this is exactly the thing that was fixed in #5011?
